sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -902,152 +903,191 @@ case class Cast(
   }
 
   // LongConverter
-  private[this] def castToLong(from: DataType): Any => Any = from match {
-    case StringType if ansiEnabled =>
-      buildCast[UTF8String](_, v => UTF8StringUtils.toLongExact(v, getContextOrNull()))
-    case StringType =>
-      val result = new LongWrapper()
-      buildCast[UTF8String](_, s => if (s.toLong(result)) result.value else null)
-    case BooleanType =>
-      buildCast[Boolean](_, b => if (b) 1L else 0L)
-    case DateType =>
-      buildCast[Int](_, d => null)
-    case TimestampType =>
-      buildCast[Long](_, t => timestampToLong(t))
-    case x: NumericType if ansiEnabled =>
-      b => x.exactNumeric.asInstanceOf[Numeric[Any]].toLong(b)
-    case x: NumericType =>
-      b => x.numeric.asInstanceOf[Numeric[Any]].toLong(b)
-    case x: DayTimeIntervalType =>
-      buildCast[Long](_, i => dayTimeIntervalToLong(i, x.startField, x.endField))
-    case x: YearMonthIntervalType =>
-      buildCast[Int](_, i => yearMonthIntervalToInt(i, x.startField, x.endField).toLong)
+  private[this] def castToLong(from: DataType): Any => Any = {
+    var exactNumeric: Numeric[Any] = null
+    var numeric: Numeric[Any] = null
```

Comment:
this is really ugly... I'm wondering if we should just remove `PhysicalNumericType`, and make `numeric` a field in the top level `PhysicalDataType`, If a type doesn't support numeric, it should throw an exception.

Then here we can just write
```
lazy val numeric = PhysicalType.numeric(from)
from match {
  ...
}
```

Reply:
This won't work because there is a `PhysicalDataType(dt).asInstanceOf[PhysicalNumericType]` thus the code will break whenever the type is not numeric type. We will see exception like `xxx cannot be cast to PhysicalNumericType`.

Reply:
Or we keep `PhysicalNumericType`, but its `numeric` takes `DataType` and throws an exception if the input is not `NumericType`. Then we can still simplify the code here.

Reply:
Code updated for simplification.